miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/display/RealMiniAppDisplay.kt

```diff
@@ -69,10 +94,29 @@ internal class RealMiniAppDisplay(
 }
 
 @VisibleForTesting
-internal class MiniAppWebViewClient(private val loader: WebViewAssetLoader) : WebViewClient() {
+internal class MiniAppWebViewClient(
+    context: Context,
+    private val loader: WebViewAssetLoader
+) : WebViewClient() {
+    @Suppress("TooGenericExceptionCaught", "SwallowedException")
+    private val bridgeJs = try {
+        val inputStream = context.assets.open("bridge.js")
+        inputStream.bufferedReader().use(BufferedReader::readText)
+    } catch (e: Exception) {
+        null
+    }
 
     override fun shouldInterceptRequest(
         view: WebView,
         request: WebResourceRequest
     ): WebResourceResponse? = loader.shouldInterceptRequest(request.url)
+
+    override fun onPageStarted(webView: WebView, url: String?, favicon: Bitmap?) {
```

Comment:
Did you get this working for Android 9?

Reply:
It's working on my Huawei device but not Android 9 emulator.

Reply:
Hmm, if it's not working in the emulator, I think we can't rely on this to be a good solution unfortunately. 

I guess we could use `onPageFinished`, but then we would need to add some kind of callback in the JS SDK to notify when the SDK is ready to use. And I don't really like that solution because then a Mini App will have to wait for some time before they can use any of the SDK functionality.

I was thinking, we could inject a `<script>` tag into the `<head>` of the HTML. But it's kind of hacky.... Something like this:
```
    override fun shouldInterceptRequest(
        view: WebView,
        request: WebResourceRequest
    ): WebResourceResponse? {
        if (request.url.host == domain && request.url.path?.endsWith("html") == true) {
            val response = loader.shouldInterceptRequest(request.url)

            val data = response?.data?.bufferedReader().use {
                it?.readText()
            }

            response?.data = data?.replace("<head>", "<head>\n<script>$bridgeJs</script>\n")?.byteInputStream()
            return response
        }

        return loader.shouldInterceptRequest(request.url)
    }
```

What do you think? It looks like our options are either:
1. Inject the JS after the page has loaded and add some `onMiniAppReady` callback in the JS SDK. 
2. or Modify the HTML to inject JS.

Reply:
Ah, it looks like it's also possible to use `onLoadResource`: https://stackoverflow.com/a/58932747

But if we use that, it would mean that the Mini App must always load at least one resource or else the JS won't get injected. I think the Mini App should always at least try to load the `favicon.ico`, but it is possible to disable that request from the client side though it's not very common to do it.

Reply:
Here is my thought of your ideas:

1/ The callback for notifying the ready of sdk could be done in `OnPageStarted` or `OnPageFinished` depending on a webview. I think we could improve it by having a function to check whether `bridge.js` is injected. 
`OnPageStarted`: do injection -> `OnPageFinished`: check if it's not injected, do injection again.
However, this requires so much work on the js sdk side.

2/ Modify the HTML to inject JS: The whole file `bridge.js` will be visible from the web inspector and yeah, it's so hacky.

3/ It works perfectly. I think we can go with this solution if we can force the miniapps load resource or just simply have it in our document declaring that favicon.ico request is mandatory.

Reply:
Okay, probably 3 is okay. We just have to make sure Mini Apps are aware of the restriction that they need to make the favicon request.